src/Fields/field.jl

```diff
@@ -36,84 +36,90 @@ A type describing the location at the face of a grid cell.
+with field boundary conditions `bcs`.
 """
-CellField(arch::AbstractArchitecture, grid, data=zeros(arch, grid)) =
-    Field(Cell, Cell, Cell, arch, grid, data)
+CellField(arch::AbstractArchitecture, grid, bcs, data=zeros(arch, grid)) =
+    Field(Cell, Cell, Cell, arch, grid, bcs, data)
 
 """
-    XFaceField(arch::AbstractArchitecture, grid, [data=zeros(arch, grid)])
+    XFaceField(arch::AbstractArchitecture, grid, bcs, [data=zeros(arch, grid)])
 
-Return a `Field{Face, Cell, Cell}` on architecture `arch` and `grid` containing `data`.
+Return a `Field{Face, Cell, Cell}` on architecture `arch` and `grid` containing `data`
+with field boundary conditions `bcs`.
 """
-XFaceField(arch::AbstractArchitecture, grid, data=zeros(arch, grid)) =
-    Field(Face, Cell, Cell, arch, grid, data)
+XFaceField(arch::AbstractArchitecture, grid, bcs, data=zeros(arch, grid)) =
+    Field(Face, Cell, Cell, arch, grid, bcs, data)
 
 """
-    YFaceField(arch::AbstractArchitecture, grid, [data=zeros(arch, grid)])
+    YFaceField(arch::AbstractArchitecture, grid, bcs, [data=zeros(arch, grid)])
 
-Return a `Field{Cell, Face, Cell}` on architecture `arch` and `grid` containing `data`.
+Return a `Field{Cell, Face, Cell}` on architecture `arch` and `grid` containing `data`
+with field boundary conditions `bcs`.
 """
-YFaceField(arch::AbstractArchitecture, grid, data=zeros(arch, grid)) =
-    Field(Cell, Face, Cell, arch, grid, data)
+YFaceField(arch::AbstractArchitecture, grid, bcs, data=zeros(arch, grid)) =
+    Field(Cell, Face, Cell, arch, grid, bcs, data)
 
 """
-    ZFaceField(arch::AbstractArchitecture, grid, [data=zeros(arch, grid)])
+    ZFaceField(arch::AbstractArchitecture, grid, bcs, [data=zeros(arch, grid)])
 
-Return a `Field{Cell, Cell, Face}` on architecture `arch` and `grid` containing `data`.
+Return a `Field{Cell, Cell, Face}` on architecture `arch` and `grid` containing `data`
+with field boundary conditions `bcs`.
 """
-ZFaceField(arch::AbstractArchitecture, grid, data=zeros(arch, grid)) =
-    Field(Cell, Cell, Face, arch, grid, data)
+ZFaceField(arch::AbstractArchitecture, grid, bcs, data=zeros(arch, grid)) =
+    Field(Cell, Cell, Face, arch, grid, bcs, data)
 
- CellField(T::DataType, arch, grid) = Field(Cell, Cell, Cell, arch, grid, zeros(T, arch, grid))
-XFaceField(T::DataType, arch, grid) = Field(Face, Cell, Cell, arch, grid, zeros(T, arch, grid))
-YFaceField(T::DataType, arch, grid) = Field(Cell, Face, Cell, arch, grid, zeros(T, arch, grid))
-ZFaceField(T::DataType, arch, grid) = Field(Cell, Cell, Face, arch, grid, zeros(T, arch, grid))
+ CellField(FT::DataType, arch, grid, bcs) =  CellField(arch, grid, bcs, zeros(FT, arch, grid))
+XFaceField(FT::DataType, arch, grid, bcs) = XFaceField(arch, grid, bcs, zeros(FT, arch, grid))
+YFaceField(FT::DataType, arch, grid, bcs) = YFaceField(arch, grid, bcs, zeros(FT, arch, grid))
+ZFaceField(FT::DataType, arch, grid, bcs) = ZFaceField(arch, grid, bcs, zeros(FT, arch, grid))
 
```

Comment:
these constructors should be extended to default bcs derived from the topology of grid:

```julia
CellField(FT::DataType, arch, grid, bcs=TracerBoundaryConditions(grid)) = ...
```

et cetera.

Reply:
Yeah I was thinking of doing that but it's not strictly correct so I decided against it. A `CellField` could require `TracerBoundaryConditions`, `PressureBoundaryConditions`, or `DiffusivityBoundaryConditions` (right now they're all the same but that might not be true in the future). Even an `XFaceField` could be a velocity field or a field holding some flux.

I'm happy to extend the constructors for convenience. We can always refactor them when a future design change requires it.

Reply:
I don't think there is a correctness problem here. We provide a sensible default for convenience; and in the case that the default is incorrect, non-default, correct boundary conditions can be specified. For example, `zeros(10)` produces an array of `Float64`, which is often what users want and is therefore convenient. This is not *always* what users want, of course, so it is possible to write `zeros(Float32, 10)` if you need to. 

Note that tracers, pressures, and diffusivities do in fact share boundary conditions (right?) so the default is correct for all of them. I think it could be annoying to have to specify boundary conditions on a field that is, for example, intended only as scratch space for diagnostics calculations. For that application having a default is nice.

Reply:
That's a good point. Will change.